TLM/TLM/Manager/Impl/RoutingManager.cs

```diff
@@ -441,89 +441,131 @@ public class RoutingManager : AbstractGeometryObservingManager, IRoutingManager
-
-
-				uint nextLaneId = nextFirstLaneId;
-				byte nextLaneIndex = 0;
-				//ushort compatibleLaneIndicesMask = 0;
-
-				while (nextLaneIndex < nextSegmentInfo.m_lanes.Length && nextLaneId != 0u) {
-					// determine valid lanes based on lane arrows
-					NetInfo.Lane nextLaneInfo = nextSegmentInfo.m_lanes[nextLaneIndex];
+                NetInfo.Direction nextDir = isNextStartNodeOfNextSegment ? NetInfo.Direction.Backward : NetInfo.Direction.Forward;
+                NetInfo.Direction nextDir2 = !nextSegIsInverted ? nextDir : NetInfo.InvertDirection(nextDir);
+
+                LaneTransitionData[] nextRelaxedTransitionDatas = null;
+                byte numNextRelaxedTransitionDatas = 0;
+                LaneTransitionData[] nextCompatibleTransitionDatas = null;
+                int[] nextCompatibleOuterSimilarIndices = null;
+                byte numNextCompatibleTransitionDatas = 0;
+                LaneTransitionData[] nextLaneConnectionTransitionDatas = null;
+                byte numNextLaneConnectionTransitionDatas = 0;
+                LaneTransitionData[] nextForcedTransitionDatas = null;
+                byte numNextForcedTransitionDatas = 0;
+                int[] nextCompatibleTransitionDataIndices = null;
+                byte numNextCompatibleTransitionDataIndices = 0;
+                int[] compatibleLaneIndexToLaneConnectionIndex = null;
+
+                if (isNextSegmentValid) {
+                    nextRelaxedTransitionDatas = new LaneTransitionData[MAX_NUM_TRANSITIONS];
+                    nextCompatibleTransitionDatas = new LaneTransitionData[MAX_NUM_TRANSITIONS];
+                    nextLaneConnectionTransitionDatas = new LaneTransitionData[MAX_NUM_TRANSITIONS];
+                    nextForcedTransitionDatas = new LaneTransitionData[MAX_NUM_TRANSITIONS];
+                    nextCompatibleOuterSimilarIndices = new int[MAX_NUM_TRANSITIONS];
+                    nextCompatibleTransitionDataIndices = new int[MAX_NUM_TRANSITIONS];
+                    compatibleLaneIndexToLaneConnectionIndex = new int[MAX_NUM_TRANSITIONS];
+                }
+
+
+                uint nextLaneId = nextFirstLaneId;
+                byte nextLaneIndex = 0;
+                //ushort compatibleLaneIndicesMask = 0;
+                int[] laneGroups = new int[nextSegmentInfo.m_lanes.Length];
+                int prevLaneGroup = 0;
+                List<NetInfo.Lane> prevSortedLanes = prevSegmentInfo.m_lanes.Where(l => (l.m_laneType & LaneConnectionManager.LANE_TYPES) != NetInfo.LaneType.None
+                    && (l.m_vehicleType & LaneConnectionManager.VEHICLE_TYPES) != VehicleInfo.VehicleType.None).OrderBy(l => l.m_position).ToList();
+                List<NetInfo.Lane> sortedLanes = nextSegmentInfo.m_lanes.Where(l => (l.m_laneType & LaneConnectionManager.LANE_TYPES) != NetInfo.LaneType.None
+                    && (l.m_vehicleType & LaneConnectionManager.VEHICLE_TYPES) != VehicleInfo.VehicleType.None).OrderBy(l => l.m_position).ToList();
+
```

Comment:
Issues:
- Linq/Collections have terrible performance on Mac. We should avoid using them whenever it's possible.
- Duplicate code (refactor to method `sortLanesBy...`?)
- Lanes are already getting sorted by outer similar index (using Linq ^^) on line 751. Can we reuse this calculation / move the logic there?
![grafik](https://user-images.githubusercontent.com/4333935/55268755-2ff59200-528d-11e9-8b98-6b91df5ee19c.png)
- Note: Both lists are sorted by `m_position` but that does not make them comparable. `m_position` cannot be used to compare lanes from different segments.
- Why do we store `NetInfo.Lane` here? lane index might be sufficient.